### DISCUSSION TITLE
osu! Mania 3D mod

### DISCUSSION THREAD

**Discussion by CdrSonan** [ppy/osu#30217] [Ideas] [answered]
I have had an idea for a "3d" mod for osu! Mania that effectively turns the playfield into a 3d-type where notes fly in towards the camera, compared to the default layout where they come in from the top or bottom.
Basically this type of layout:
![sdvx](https://github.com/user-attachments/assets/ca931d5f-34d4-4cda-a239-6f37f38b08e9)

Looking at the discussions, this has been brought up two times before, in https://github.com/ppy/osu/discussions/25447 and https://github.com/ppy/osu/discussions/27204 . However, the first hasn't seen any updates in almost a year, and the second only got a single reply pointing to the first discussion.

I would like to work on this mod right now, but wanted to check in here beforehand instead of just opening a PR, to make sure I'm not missing anything important.

My implementation would mainly consist of shear transforms being applied to all notes and the playfield to make them look as if they're on a 3d "ground plane", and a modulation of the scroll speed as notes fly towards the camera to get a matching parallax effect. This may or may not require a custom easing function, but either way, I am confident I can figure out the math to make the perspective make sense and have everything behave correctly.

The UI would remain as-is, without any additional transforms.

The main challenge would be figuring out when notes appear and scroll speed. In typical 3d scroller rhythm games, notes become visible a lot earlier than in their 2d counterparts, and scroll speed doesn't translate between them 1:1 (speed in a 3d world space vs in a 2d screen space)

Another consideration is the angle at which the camera "looks down" onto the playfield, which corresponds to how high up on the screen the playfield ends. (look at the image for reference). Some rhythm games, like Sound Voltex, change this angle over the course of a song for artistic effect, but for this mod, I'd simply make it a mod setting, like the Double Time speed multiplier.

What do you think about this? And is it okay if I work on this, in case there is ongoing development for it already?

**Comment by bdach** [comment, marked as answer]
> And is it okay if I work on this, in case there is ongoing development for it already?

https://github.com/ppy/osu-framework/pull/6062 is a hard prerequisite for this. Shear is not enough.

**Reply by CdrSonan** [reply]
Sorry. You are absolutely right.
Looks like I misunderstood the shear transform.

Anyway, in that case I'm going to work on something else and revisit this once the framework supports it.
